src/analysis/skill_curve.rs

```diff
@@ -0,0 +1,223 @@
+                    Err(right_index) => {
+                        // Saturate both left and right index at the bounds of the
+                        // array. Both could go out of bounds.
+                        let left_index = right_index.saturating_sub(1);
+                        // This assumes len can never be 0. This only works out
+                        // for us, because we truncate all the empty segments away.
+                        let right_index = right_index.min(weighted_segment_times.len() - 1);
+
+                        if left_index == right_index {
+                            weighted_segment_times[left_index].1
+                        } else {
+                            let right = weighted_segment_times[right_index];
+                            let left = weighted_segment_times[left_index];
+
+                            interpolate(percentile, left, right)
+                        }
+                    }
+                }
+            })
+    }
+
+    /// This function returns an iterator that iterates over each segment and
+    /// yields their split times for the percentile specified. A percentile of 0
+    /// yields the fastest times, while a percentile of 1 yields the slowest
+    /// times. The offset provided is the initial split time going into the
+    /// first segment. This is only relevant when the segments don't represent
+    /// the beginning of the run.
+    pub fn iter_split_times_at_percentile(
+        &self,
+        percentile: f64,
+        offset: TimeSpan,
+    ) -> impl Iterator<Item = TimeSpan> + '_ {
+        let mut sum = offset;
+        self.iter_segment_times_at_percentile(percentile)
+            .map(move |segment_time| {
+                sum += segment_time;
+                sum
+            })
+    }
+
+    /// Searches the curve for the final run time specified and returns the
+    /// percentile the time can be found at. The percentile is always within the
+    /// range 0..1. If the time specified can not be found, the percentile
+    /// saturates at one of those boundaries.
+    pub fn find_percentile_for_time(&self, offset: TimeSpan, time_to_find: TimeSpan) -> f64 {
+        let (mut perc_min, mut perc_max) = (0.0, 1.0);
+
+        // Try to find the correct percentile
+        for _ in 0..TRIES {
+            let percentile = (perc_max + perc_min) / 2.0;
+
+            let sum = self
+                .iter_segment_times_at_percentile(percentile)
+                .fold(offset, |sum, segment_time| sum + segment_time);
+
+            // Binary search the correct percentile
+            match sum.partial_cmp(&time_to_find) {
+                Some(Ordering::Equal) => return percentile,
+                Some(Ordering::Less) => perc_min = percentile,
+                Some(Ordering::Greater) => perc_max = percentile,
+                None => {}
+            }
+        }
+
+        (perc_max + perc_min) / 2.0
```

Comment:
Should we be distinguishing between exact and inexact matches here like we do for `iter_segment_times_at_percentile`?

Reply:
Here we could signal to the outside that it's not an exact match, but idk if we really need one. 50 tries of binary search yields an accuracy of 50 bits, which is close to the 64-bits of accuracy f64 could give us. So if we really cared for full accuracy, we could just do 64 rounds.